pkg/controller/restore_controller.go

```diff
@@ -462,43 +468,92 @@ func mostRecentCompletedBackup(backups []*api.Backup) *api.Backup {
 	return nil
 }
 
-func (c *restoreController) fetchBackup(objectStore cloudprovider.ObjectStore, name string) (*api.Backup, error) {
-	backup, err := c.backupLister.Backups(c.namespace).Get(name)
-	if err == nil {
-		return backup, nil
+// fetchBackupInfo checks the backup lister for a backup that matches the given name. If it doesn't
+// find it, it tries to retrieve it from one of the backup storage locations.
+func (c *restoreController) fetchBackupInfo(backupName string, pluginManager plugin.Manager) (backupInfo, error) {
+	var info backupInfo
+	var err error
+	info.backup, err = c.backupLister.Backups(c.namespace).Get(backupName)
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return backupInfo{}, errors.WithStack(err)
+		}
+
+		logContext := c.logger.WithField("backupName", backupName)
+		logContext.Debug("Backup not found in backupLister, checking each backup location directly, starting with default...")
+		return c.fetchFromBackupStorage(backupName, pluginManager)
 	}
 
-	if !apierrors.IsNotFound(err) {
-		return nil, errors.WithStack(err)
+	location, err := c.backupLocationLister.BackupStorageLocations(c.namespace).Get(info.backup.Spec.StorageLocation)
+	if err != nil {
+		return backupInfo{}, errors.WithStack(err)
 	}
 
-	logContext := c.logger.WithField("backupName", name)
+	info.objectStore, err = getObjectStoreForLocation(location, pluginManager)
+	if err != nil {
+		return backupInfo{}, errors.Wrap(err, "error initializing object store")
+	}
+	info.bucketName = location.Spec.ObjectStorage.Bucket
+
+	return info, nil
+}
 
-	logContext.Debug("Backup not found in backupLister, checking object storage directly")
-	backup, err = c.getBackup(objectStore, c.bucket, name)
+// fetchFromBackupStorage checks each backup storage location, starting with the default,
+// looking for a backup that matches the given backup name.
+func (c *restoreController) fetchFromBackupStorage(backupName string, pluginManager plugin.Manager) (backupInfo, error) {
+	locations, err := c.backupLocationLister.BackupStorageLocations(c.namespace).List(labels.Everything())
 	if err != nil {
-		return nil, err
+		return backupInfo{}, err
+	}
+
+	idx := 1
+	orderedLocations := make([]*api.BackupStorageLocation, len(locations))
+	for i := 0; i < len(locations); i++ {
```

Comment:
One issue with this implementation is that if for some reason, there is no backup storage location with the default name, then `orderedLocations[0]` will never get filled in and we'll eventually get a panic (I think).  One alternative way of writing this for your consideration:
```
func orderedBackupLocations(locations []*api.BackupStorageLocation, defaultLocationName string) []*api.BackupStorageLocation {
	var result []*api.BackupStorageLocation

	for i := range locations {
		if locations[i].Name == defaultLocationName {
			// put the default location first
			result = append(result, locations[i])
			// append everything before the default
			result = append(result, locations[:i]...)
			// append everything after the default
			result = append(result, locations[i+1:]...)

			return result
		}
	}

	return locations
}
```

Reply:
Yep. Very good point!

Reply:
This is a lot more elegant 👍